src/hotspot/share/gc/g1/g1DirtyCardQueue.cpp

```diff
@@ -237,11 +130,49 @@ void G1DirtyCardQueueSet::enqueue_completed_buffer(BufferNode* cbn) {
   }
 }
 
+// Thread-safe attempt to remove and return the first buffer from
+// the _completed queue, using the LockFreeQueue::try_pop() underneath.
+// It has a restriction that it may return NULL when there are objects
+// in the queue if there is a concurrent push/append operation.
+BufferNode* G1DirtyCardQueueSet::dequeue_completed_buffer() {
+  using Status = LockFreeQueuePopStatus;
+  Thread* current_thread = Thread::current();
+  while (true) {
+    // Use GlobalCounter critical section to avoid ABA problem.
+    // The release of a buffer to its allocator's free list uses
+    // GlobalCounter::write_synchronize() to coordinate with this
+    // dequeuing operation.
+    // We use a CS per iteration, rather than over the whole loop,
+    // because we're not guaranteed to make progress. Lingering in
+    // one CS could defer releasing buffer to the free list for reuse,
+    // leading to excessive allocations.
+    GlobalCounter::CriticalSection cs(current_thread);
+    Pair<Status, BufferNode*> pop_result = _completed.try_pop();
+    switch (pop_result.first) {
+      case Status::success:
+        return pop_result.second;
+      case Status::operation_in_progress:
+        // This could happen when a concurrent operation interferes with
```

Comment:
I think most of this comment doesn't belong here.  What's here now is largely restating what should be the documentation for `operation_in_progress`.